### PR TITLE
Merkle tree version for arrays with length power of 2. 

### PR DESCRIPTION
**Pull request by isoprophlex**
I added a first version of a Merkle Tree implementation based on an Array but only with an input array with length power of two. 

I thought of doing it without an struct, but with it, the code is nicer with it.

### CODE REVIEW

**Comment by gianbelinche** [comment]
Comment for following commits, they should use imperative (https://cbea.ms/git-commit/#imperative point 5)

Fix previous test instead of Fixed previous test

File: rusty-merkle-tree/src/merkle_tree.rs [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+pub enum ArrayError{
+    Empty,
+    NotPowerOfTwo
+}
+#[derive(Debug, PartialEq, Clone)]
+pub struct MerkleTree {
+    input: Vec<String>,
+    leaves: Vec<String>
+}
+const SMALL_ARRAY: usize = 1;
```
Comment (gianbelinche) [inline rusty-merkle-tree/src/merkle_tree.rs]: warning: constant `SMALL_ARRAY` is never used

File: rusty-merkle-tree/src/merkle_tree.rs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+impl MerkleTree {
+    //  This function creates the MerkleTree struct
+    // verying the input array
+    fn new(leaves_array: Vec<String>) -> Result<MerkleTree, ArrayError> {
+        if let Err(error) = MerkleTree::verify_input(&leaves_array) {
+            return Err(error);
+        }
```
Comment (gianbelinche) [inline rusty-merkle-tree/src/merkle_tree.rs]: You can use ? here

MerkleTree::verify_input(&leaves_array)?

File: rusty-merkle-tree/src/merkle_tree.rs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+            return Err(ArrayError::Empty);
+        }
+        if !is_power_of_two(array.len()) {
+            return Err(ArrayError::NotPowerOfTwo);
+        }
+        Ok(())
+    }
+    fn calculate_merkle_root(&self) -> String {
+        if self.leaves.is_empty() {
+            return String::new(); // Devolver una cadena vacía si no hay hojas
```
Comment (gianbelinche) [inline rusty-merkle-tree/src/merkle_tree.rs]: Comment in english (There are others too)

**Review by gianbelinche — CHANGES_REQUESTED** [review]
Fix the changes commented, everything else is fine

**Review by gianbelinche — COMMENTED** [review]
Its probably not a good idea to add new commits to an open PR, either maintain it on draft or let the reviewer know

**Comment by gianbelinche** [comment]
I forgot to add this warning

warning: associated items `new`, `hash_leaves`, `verify_input`, `calculate_merkle_root`, and `hash_nodes` are never used

src/merkle_tree.rs:18:8

File: rusty-merkle-tree/src/merkle_tree.rs (hunk 2 above)
Comment (isoprophlex) [reply rusty-merkle-tree/src/merkle_tree.rs]: It returns Result<(), ArrayError>, how can I apply that here?

Comment (gianbelinche) [reply rusty-merkle-tree/src/merkle_tree.rs]: Both approaches do the same thing, the ? is more concise

**Review by gianbelinche — APPROVED** [review]
Everything looks fine now!

**Pull request merged**